lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -136,15 +146,246 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
+  MLOG(MINFO) << "During release state of session changed to "
+               << session_fsm_state_to_str(session->get_state())
+	       << " Current version is " << session->get_current_version();
+
+  /* Call for all rules to be de-associated from session 
+   * Message to UPF need to be filled from session i.e. of session's
+   * current_version_ 
+   * current_state_ and
+   * dnn can directly be used.
+   */
+  m5g_remove_rules_from_session_for_termination(imsi,
+		   session, dnn, session_uc);
+
+  /* Forcefully terminate session context on time out
+   * time out = 5000ms from sessiond.yml config file
+   */
+  MLOG(MINFO) << "Scheduling a force termination timeout for session_id "
+	      << session_id << " in " 
+	      << session_force_termination_timeout_ms_ << "ms";
+
+  evb_->runAfterDelay(
+     [this, imsi, session_id] {
+        m5g_handle_termination_on_timeout(imsi, session_id);
+     }, session_force_termination_timeout_ms_);
+}
+
+/*Function to handle termination if UPF doesn't send required report
+ * As per current implementation, upf report is not in place and 
+ * termination on time out will be executed forcefully
+ */
+void SessionStateEnforcer::m5g_handle_termination_on_timeout(
+      const std::string& imsi, const std::string& session_id)
+{
+  auto session_map    = session_store_.read_sessions_for_deletion({imsi});
+  auto session_update = SessionStore::get_default_session_update(session_map);
+  bool marked_termination =
+      session_update[imsi].find(session_id) != session_update[imsi].end();
+  MLOG(MINFO) << "Forced termination timeout! Checking if termination has to "
+               << "be forced for " << session_id << "... => "
+               << (marked_termination ? "YES" : "NO");
+  /* If the session doesn't exist in the session_update, then the session was
+   * already released and terminated
+   */
+  if (marked_termination) {
+     /*call to remove session from map*/
+     m5g_complete_termination(session_map, imsi, session_id, session_update);
+
+     bool update_success = session_store_.update_sessions(session_update);
+     if (update_success) {
+       MLOG(MINFO) << "Updated session termination of " << session_id
+                   << " in to SessionStore";
+     } else {
+       MLOG(MERROR) << "Failed to update session termination of " << session_id
+                    << " in to SessionStore";
+     }
+   } else {
+       MLOG(MERROR) << "Nothing to remove as no respective entry found for "
+	            << "session id " << session_id 
+	 	    << " of IMSI " << imsi;
+   }
+}
+
+/*Function will clean up all resources related to requested session
+ * if it is last session entry, then delete the imsi
+ * This function can be invoked from 2 different sources
+ * 1. Time out and forcefully terminates session
+ * 2. Once UPF sends report to SessionD
+ * The 2nd one we are not taking care now.
+ */
+void SessionStateEnforcer::m5g_complete_termination(

```

Comment:
Do we notify the access on only when we transition into the RELEASED state? Or do we also propagate the state when the session has been terminated as well?

Reply:
Take care.